### ISSUE TITLE
CCMake Issues Building from Source

### ISSUE THREAD

**Issue by nickjbrowning** [openmm/openmm#3106]
Just a small set of issues when building the Makefile that I can't seem to figure out. I'm using the HPC Developer libraries obtained from here: https://developer.nvidia.com/hpc

compiling from source gives the following error when using ccmake:
```

CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
 Please set them or make sure they are set and tested correctly in the CMake files:
 CUDA_cufft_LIBRARY (ADVANCED)
     linked by target "OpenMMCUDA" in directory /home/nick/git/openmm/platforms/cuda/sharedTarget

```

I've checked the "advanced" mode, and it seems like there are quite a few issues with finding various CUDA libraries, including:

```
CUDA_cublas_LIBRARY              CUDA_cublas_LIBRARY-NOTFOUND                                                                          
 CUDA_cudadevrt_LIBRARY           /opt/nvidia/hpc_sdk/Linux_x86_64/20.11/cuda/lib64/libcudadevrt.a                                      
 CUDA_cudart_static_LIBRARY       /opt/nvidia/hpc_sdk/Linux_x86_64/20.11/cuda/lib64/libcudart_static.a                                  
 CUDA_cufft_LIBRARY               CUDA_cufft_LIBRARY-NOTFOUND                                                                           
 CUDA_cupti_LIBRARY               CUDA_cupti_LIBRARY-NOTFOUND                                                                           
 CUDA_curand_LIBRARY              CUDA_curand_LIBRARY-NOTFOUND                                                                          
 CUDA_cusolver_LIBRARY            CUDA_cusolver_LIBRARY-NOTFOUND                                                                        
 CUDA_cusparse_LIBRARY            CUDA_cusparse_LIBRARY-NOTFOUND                                                                        
 CUDA_nppc_LIBRARY                CUDA_nppc_LIBRARY-NOTFOUND                                                                            
 CUDA_nppial_LIBRARY              CUDA_nppial_LIBRARY-NOTFOUND                                                                          
 CUDA_nppicc_LIBRARY              CUDA_nppicc_LIBRARY-NOTFOUND                                                                          
 CUDA_nppicom_LIBRARY             CUDA_nppicom_LIBRARY-NOTFOUND                                                                         
 CUDA_nppidei_LIBRARY             CUDA_nppidei_LIBRARY-NOTFOUND
 CUDA_nppif_LIBRARY               CUDA_nppif_LIBRARY-NOTFOUND                                                                           
 CUDA_nppig_LIBRARY               CUDA_nppig_LIBRARY-NOTFOUND                                                                           
 CUDA_nppim_LIBRARY               CUDA_nppim_LIBRARY-NOTFOUND                                                                           
 CUDA_nppist_LIBRARY              CUDA_nppist_LIBRARY-NOTFOUND                                                                          
 CUDA_nppisu_LIBRARY              CUDA_nppisu_LIBRARY-NOTFOUND                                                                          
 CUDA_nppitc_LIBRARY              CUDA_nppitc_LIBRARY-NOTFOUND                                                                          
 CUDA_npps_LIBRARY                CUDA_npps_LIBRARY-NOTFOUND                                                                            
 CUDA_nvToolsExt_LIBRARY          /opt/nvidia/hpc_sdk/Linux_x86_64/20.11/cuda/lib64/libnvToolsExt.so                                    
 CUDA_nvrtc_LIBRARY               CUDA_nvrtc_LIBRARY-NOTFOUND
```

I've set ```CUDA_HOME``` to ```/opt/nvidia/hpc_sdk/Linux_x86_64/20.11/cuda```, as well as appended ```/opt/nvidia/hpc_sdk/Linux_x86_64/20.11/math_libs/lib64``` and ```/opt/nvidia/hpc_sdk/Linux_x86_64/20.11/cuda/lib64``` to ```LD_LIBRARY_PATH``` such that the linker can find the relevant math libraries/cuda libraries, however as you can see, ccmake has trouble for many of these (they exist!).

Even if I manually enter all of these in and reconfigure, it still shows that they're not found.

Has anyone else ran into similar issues?

**Comment by peastman** [comment]
In CMake, set `CUDA_TOOLKIT_ROOT_DIR` to point to the CUDA toolkit.

I'm not familiar with the HPC SDK, but perhaps it includes the toolkit?  I've always used the installers from https://developer.nvidia.com/cuda-toolkit.

**Comment by nickjbrowning** [comment]
@peastman 

thanks for the quick reply. I'm using the HPC SDK since it also contains the pgifortran compilers + CUDA fortran wrappers that I need for another project.

reverting back to the default nvidia cuda SDK does indeed solve the issue, though!

**Comment by dmclark17** [comment]
Hi @peastman, I was able to reproduce this issue and will investigate further. Should I create a new issue, or can we re-open this one for tracking?

**Comment by peastman** [comment]
If it's the same issue, keeping it here makes sense.  I'll reopen this.

**Comment by dmclark17** [comment]
I was able to compile OpenMM with the NVIDIA HPC SDK; however, I did switch to using the [FindCUDAToolkit](https://cmake.org/cmake/help/git-stage/module/FindCUDAToolkit.html) command instead of [FindCUDA](https://cmake.org/cmake/help/latest/module/FindCUDA.html). 

FindCUDAToolkit was introduced in cmake 3.17 and supersedes the deprecated FindCUDA command. @peastman, would it be possible to make this switch in OpenMM? I understand that it is a pretty significant bump in minimal cmake version.

**Comment by peastman** [comment]
It looks like 3.17 was released over a year ago.  I think that's old enough that's it reasonable to require it.  Conda-forge is distributing 3.20.

**Comment by dmclark17** [comment]
Great! I'll prepare a pull request with the cmake change. 

Another issue that I came across was, during the `make PythonInstall` step, the C compiler seemed to be called instead of the C++ one. For example, when I compile with gcc/g++, I see:
```
gcc -pthread -B /home/davidclark/miniconda3/compiler_compat -Wl,--sysroot=/ -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes -fPIC -DMAJOR_VERSION=7 -DMINOR_VERSION=5 ...(include pathes)... -c src/swig_doxygen/OpenMMSwig.cxx -o build/temp.linux-x86_64-3.8/src/swig_doxygen/OpenMMSwig.o -std=c++11
```

This caused an issue when using the `nvc` host compiler; however, running `CC=$CXX make PythonInstall` resolved the issue. I tried explicitly setting the language parameter in the [Extension object](https://docs.python.org/3/distutils/apiref.html#distutils.core.Extension) in setup.py [here](https://github.com/openmm/openmm/blob/a29596c70119ffc5ddbd81e46472cf80989f2137/wrappers/python/setup.py#L219), but it didn't change the compiler being used. The workaround is pretty simply, but I was wondering if you knew why the C compiler is called. Thanks!

**Comment by peastman** [comment]
I don't know.  We could try adding the option `"language": "c++"` to the Extension's arguments:

https://github.com/openmm/openmm/blob/a29596c70119ffc5ddbd81e46472cf80989f2137/wrappers/python/setup.py#L219-L226

But that shouldn't be necessary.  It's supposed to [automatically detect it](https://docs.python.org/3/distutils/apiref.html#distutils.core.Extension) from the source file's extension, which in this case is `.cxx`.

**Comment by dmclark17** [comment]
Explicitly setting the language option did not resolve the problem—it seems like it is an issue with setuptools: https://github.com/pypa/setuptools/issues/1732. 

To compile OpenMM with the NVIDIA HPC SDK, I had to:

1. Set the following environment variables:

```
export CFLAGS="-noswitcherror"
export CXXFLAGS="-noswitcherror"
export CPATH=[install-prefix]/[arch]/[nvhpc-version]/math_libs/include:$CPATH
```

2. Add the following to OpenMM's cmake configuration:

```
CUDA_cufft_LIBRARY="[install-prefix]/[arch]/[nvhpc-version]/math_libs/lib64/libcufftw.so"
```

3. Set `nvc++` as the C compiler for the PythonInstall step:

```
CC=$CXX make PythonInstall
```

Would it be possible to add these instructions to the Compiling OpenMM from Source Code portion of the documentation?

**Comment by peastman** [comment]
To make sure I understand, these steps are needed when CUDA is installed via the HPC SDK, but *not* when it's installed via the CUDA toolkit?  If so, would it be best to just advise people to use the toolkit?  What are the advantages of using the HPC SDK instead?

**Comment by dmclark17** [comment]
Correct, these steps are only needed when using the HPC SDK and `nvc` / `nvc++` as the host compilers. 

The HPC SDK comes with various communication libraries, like NCCL and NVSHMEM, and compilers that support automatic GPU acceleration of standard language constructs. If someone is already using the HPC SDK, it may be helpful to have build notes.

**Comment by peastman** [comment]
Ok!  I've marked this as a documentation issue.  We should recommend that people generally use the toolkit, since that's easiest, but also list the steps needed if someone wants to use the HPC SDK instead.
